README.md
# notepack

[![Build Status](https://github.com/darrachequesne/notepack/workflows/CI/badge.svg?branch=main)](https://github.com/darrachequesne/notepack/actions)

A fast [Node.js](http://nodejs.org) implementation of the latest [MessagePack](http://msgpack.org) [spec](https://github.com/msgpack/msgpack/blob/master/spec.md).

## Notes

* `undefined` is encoded as `nil`
* `Date` objects are encoded following the [Timestamp extension](https://github.com/msgpack/msgpack/blob/master/spec.md#timestamp-extension-type), e.g. `new Date('2000-06-13T00:00:00.000Z')` => `<Buffer d6 ff 39 45 79 80>`
* `ArrayBuffer` are encoded as `ext 8/16/32 [0, data]`, e.g. `Uint8Array.of(1, 2, 3, 4)` => `<Buffer c7 04 00 01 02 03 04>`

## Install

```
npm install notepack.io
```

## Usage

```js
const notepack = require('notepack.io');

const encoded = notepack.encode({ foo: 'bar'}); // <Buffer 81 a3 66 6f 6f a3 62 61 72>
const decoded = notepack.decode(encoded); // { foo: 'bar' }
```

## Browser

A browser version of notepack is also available (2.0 kB minified/gzipped)

```html
<script src="https://unpkg.com/notepack.io@2.3.0/dist/notepack.min.js"></script>
<script>
  console.log(notepack.decode(notepack.encode([1, '2', new Date()])));
  // [1, "2", Thu Dec 08 2016 00:00:01 GMT+0100 (CET)]
</script>
```

## Common questions

### How to encode custom types?

This library does not currently support [extension types](https://github.com/msgpack/msgpack/blob/master/spec.md#extension-types). That being said, you can create a `toJSON()` method on your object, which will be used when encoding:

```js
class MyClass {
  toJSON() {
    return 42;
  }
}
```

### How to handle BigInt values?

You can use the `toJSON()` method:

```js
// always as string
BigInt.prototype.toJSON = function () {
  return String(this);
};

// or either as string or number, depending on the value
BigInt.prototype.toJSON = function () {
  var isSafeNumber = Number.MIN_SAFE_INTEGER <= this && this <= Number.MAX_SAFE_INTEGER;
  return isSafeNumber ? Number(this) : String(this);
};
```

### Handle to handle ES6 Set and Map values?

Again, `toJSON()` to the rescue:

```js
// convert the set to an array
// example: Set(3) { 1, 2, 3 } into [ 1, 2, 3 ]
Set.prototype.toJSON = function () {
  return [...this];
}

// convert the map to an array of array
// example: Map(2) { 1 => '2', '3' => 4 } into [ [ 1, '2' ], [ '3', 4 ] ]
Map.prototype.toJSON = function () {
  return [...this];
}
```

## Performance

Performance is currently comparable to msgpack-node (which presumably needs optimizing and suffers from JS-native overhead) and is significantly faster than other implementations. Several micro-optimizations are used to improve the performance of short string and Buffer operations.

The `./benchmarks/run` output on my machine is:

```
$ node -v
v12.15.0
$ ./benchmarks/run
Encoding (this will take a while):
+----------------------------+-------------------+-----------------+----------------+---------------+
|                            │ tiny              │ small           │ medium         │ large         |
+----------------------------+-------------------+-----------------+----------------+---------------+
| notepack                   │ 2,187,481 ops/sec │ 510,581 ops/sec │ 39,187 ops/sec │ 231 ops/sec   |
+----------------------------+-------------------+-----------------+----------------+---------------+
| msgpack-js                 │ 111,209 ops/sec   │ 95,346 ops/sec  │ 9,896 ops/sec  │ 121 ops/sec   |
+----------------------------+-------------------+-----------------+----------------+---------------+
| msgpack-lite               │ 524,993 ops/sec   │ 195,466 ops/sec │ 18,269 ops/sec │ 242 ops/sec   |
+----------------------------+-------------------+-----------------+----------------+---------------+
| @msgpack/msgpack           │ 723,885 ops/sec   │ 292,447 ops/sec │ 30,438 ops/sec │ 80.26 ops/sec |
+----------------------------+-------------------+-----------------+----------------+---------------+
| JSON.stringify (to Buffer) │ 1,359,120 ops/sec │ 335,024 ops/sec │ 15,721 ops/sec │ 25.97 ops/sec |
+----------------------------+-------------------+-----------------+----------------+---------------+
Decoding (this will take a while):
+--------------------------+-------------------+-----------------+----------------+---------------+
|                          │ tiny              │ small           │ medium         │ large         |
+--------------------------+-------------------+-----------------+----------------+---------------+
| notepack                 │ 3,165,012 ops/sec │ 642,348 ops/sec │ 32,173 ops/sec │ 249 ops/sec   |
+--------------------------+-------------------+-----------------+----------------+---------------+
| msgpack-js               │ 1,255,151 ops/sec │ 280,944 ops/sec │ 24,396 ops/sec │ 243 ops/sec   |
+--------------------------+-------------------+-----------------+----------------+---------------+
| msgpack-lite             │ 667,059 ops/sec   │ 144,927 ops/sec │ 11,922 ops/sec │ 175 ops/sec   |
+--------------------------+-------------------+-----------------+----------------+---------------+
| @msgpack/msgpack         │ 1,760,026 ops/sec │ 353,698 ops/sec │ 18,816 ops/sec │ 45.68 ops/sec |
+--------------------------+-------------------+-----------------+----------------+---------------+
| JSON.parse (from Buffer) │ 1,750,845 ops/sec │ 407,212 ops/sec │ 24,999 ops/sec │ 35.77 ops/sec |
+--------------------------+-------------------+-----------------+----------------+---------------+
* Note that JSON is provided as an indicative comparison only
```
## License

MIT


HISTORY.md
## [3.0.1](https://github.com/darrachequesne/notepack/compare/3.0.0...3.0.1) (2022-05-22)

The browser bundle was not updated in the previous release.



# [3.0.0](https://github.com/darrachequesne/notepack/compare/2.3.0...3.0.0) (2022-05-22)


### Features

* add support for BigInt.prototype.toJSON ([#28](https://github.com/darrachequesne/notepack/issues/28)) ([9ec1fcd](https://github.com/darrachequesne/notepack/commit/9ec1fcde640dfb57b69ff5f8b45ffa5439a87700))

This change allows to properly encode [BigInt](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt) objects:

```js
// always as string
BigInt.prototype.toJSON = function () {
  return String(this);
};
// or either as string or number, depending on the value
BigInt.prototype.toJSON = function () {
  var isSafeNumber = Number.MIN_SAFE_INTEGER <= this && this <= Number.MAX_SAFE_INTEGER;
  return isSafeNumber ? Number(this) : String(this);
};
```

* encode ArrayBuffer objects as bin ([8209327](https://github.com/darrachequesne/notepack/commit/8209327a69e471625f98aa1092d2f0e74e45da4f))

```js
import { encode } from "notepack.io";

encode(Uint8Array.of(1, 2, 3, 4));

// before: <Buffer c7 04 00 01 02 03 04>
// after: <Buffer c4 04 01 02 03 04>
```

ArrayBuffer objects encoded with previous versions of the library will still be decoded, but as Buffer instead of Uint8Array.

Reference: https://github.com/msgpack/msgpack/blob/master/spec.md#bin-format-family

* encode Date objects following the official timestamp extension ([9fb6275](https://github.com/darrachequesne/notepack/commit/9fb62754e826984d9eccbb72c4e9c067ab5cb227))

```js
import { encode } from "notepack.io";

encode(new Date("2000-06-13T00:00:00.000Z"));

// before: <Buffer d7 00 00 00 00 df b7 62 9c 00>
// after: <Buffer d6 ff 39 45 79 80>
```

Date objects encoded with previous versions of the library will still be properly decoded.

Reference: https://github.com/msgpack/msgpack/blob/master/spec.md#timestamp-extension-type

* match JSON.stringify() behavior ([21c6592](https://github.com/darrachequesne/notepack/commit/21c6592cf357c972387284b96999ba8273f08035))

The library will now match the behavior of the JSON.stringify() method:

- undefined is now encoded as nil (0xc0)
- undefined values in objects are now ignored

```js
import { encode, decode } from "notepack.io";

const value = {
  a: undefined,
  b: null,
  c: [undefined, null]
};

decode(encode(value));

// before:
// {
//   a: undefined,
//   b: null,
//   c: [undefined, null]
// }
// after:
// {
//   b: null,
//   c: [null, null]
// }
```

# [2.3.0](https://github.com/darrachequesne/notepack/compare/2.2.0...v2.3.0) (2020-03-15)


### Performance Improvements

* **decode:** add a cache for buffer-to-string conversions ([3c0e5a6](https://github.com/darrachequesne/notepack/commit/3c0e5a66332e50ce31749f0159a533156edbdd3d))
* **encode:** add a cache for string-to-buffer conversions ([60e8b0b](https://github.com/darrachequesne/notepack/commit/60e8b0b4b16b05e702334fe731df1ec43d1a9f14))



# [2.2.0](https://github.com/darrachequesne/notepack/compare/2.1.3...2.2.0) (2018-12-18)



<a name="2.1.3"></a>
## [2.1.3](https://github.com/darrachequesne/notepack/compare/2.1.2...2.1.3) (2018-05-14)


### Bug Fixes

* **browser:** fix utf-8 decoder ([#16](https://github.com/darrachequesne/notepack/issues/16)) ([abbf3a5](https://github.com/darrachequesne/notepack/commit/abbf3a5))



<a name="2.1.2"></a>
## [2.1.2](https://github.com/darrachequesne/notepack/compare/2.1.1...2.1.2) (2017-08-23)


### Bug Fixes

* **encode:** remove the unsafe integer check ([#15](https://github.com/darrachequesne/notepack/issues/15)) ([bb8140c](https://github.com/darrachequesne/notepack/commit/bb8140c))



<a name="2.1.1"></a>
## [2.1.1](https://github.com/darrachequesne/notepack/compare/2.1.0...2.1.1) (2017-08-08)


### Bug Fixes

* **browser:** fix decoding for strings with surrogate pairs ([#13](https://github.com/darrachequesne/notepack/issues/13)) ([a89e566](https://github.com/darrachequesne/notepack/commit/a89e566))
* **browser:** preserve the offset and length when creating a DataView ([#11](https://github.com/darrachequesne/notepack/issues/11)) ([bd91aa7](https://github.com/darrachequesne/notepack/commit/bd91aa7))



<a name="2.1.0"></a>
# [2.1.0](https://github.com/darrachequesne/notepack/compare/2.0.1...2.1.0) (2017-07-31)


### Features

* add support for toJSON method ([#8](https://github.com/darrachequesne/notepack/issues/8)) ([9345f9f](https://github.com/darrachequesne/notepack/commit/9345f9f)), closes [#7](https://github.com/darrachequesne/notepack/issues/7)



<a name="2.0.1"></a>
## [2.0.1](https://github.com/darrachequesne/notepack/compare/2.0.0...2.0.1) (2017-06-06)


### Bug Fixes

* **encode:** fix encoding for non-finite numbers ([#4](https://github.com/darrachequesne/notepack/issues/4)) ([f0ed0f3](https://github.com/darrachequesne/notepack/commit/f0ed0f3))



<a name="2.0.0"></a>
# [2.0.0](https://github.com/darrachequesne/notepack/compare/1.0.1...2.0.0) (2017-05-19)


### Features

* Add support for ArrayBuffer ([#2](https://github.com/darrachequesne/notepack/issues/2)) ([9eec8dc](https://github.com/darrachequesne/notepack/commit/9eec8dc))
* **browser:** switch from Buffer polyfill to ArrayBuffer ([#1](https://github.com/darrachequesne/notepack/issues/1)) ([8d7ce87](https://github.com/darrachequesne/notepack/commit/8d7ce87))


test/browser.js
'use strict';

const notepack = {
  encode: require('../browser/encode'),
  decode: require('../browser/decode')
};
const expect = require('chai').expect;

function array(length) {
  const arr = new Array(length);
  for (let i = 0; i < arr.length; i++) {
    arr[i] = 0;
  }
  return arr;
}

function map(length) {
  const result = {};
  for (let i = 0; i < length; i++) {
    result[i + ''] = 0;
  }
  return result;
}

function checkDecode(value, hex) {
  const decodedValue = notepack.decode(Buffer.from(hex, 'hex'));
  expect(decodedValue).to.deep.equal(value, 'decode failed');
}

function checkEncode(value, hex) {
  const encodedHex = Buffer.from(notepack.encode(value)).toString('hex');
  expect(encodedHex).to.equal(hex, 'encode failed');
}

function check(value, hex) {
  checkEncode(value, hex);
  checkDecode(value, hex);

  // And full circle for fun
  expect(notepack.decode(notepack.encode(value))).to.deep.equal(value);
}

describe('notepack (browser build)', function() {
  it('ArrayBuffer view', function() {
    expect(notepack.decode(Uint8Array.from([ 0x93, 1, 2, 3 ]))).to.deep.equal([ 1, 2, 3 ]);
  });

  it('offset ArrayBuffer view', function() {
    const buffer = new ArrayBuffer(14);
    const view = new Uint8Array(buffer);

    // Fill with junk before setting the encoded data
    view.fill(0xFF);

    // Put the encoded data somewhere in the middle of the buffer
    view.set([ 0x93, 1, 2, 3 ], 4);

    expect(notepack.decode(new Uint8Array(buffer, 4, 4))).to.deep.equal([ 1, 2, 3 ]);
    expect(notepack.decode(new Uint16Array(buffer, 4, 2))).to.deep.equal([ 1, 2, 3 ]);
  });

  it('toJSON', function () {
    const obj = {
      a: 'b',
      toJSON: function () {
        return 'c';
      }
    };
    expect(notepack.encode(obj)).to.deep.equal(notepack.encode('c'));
  });

  it('utf-8', function() {
    // 1-byte
    expect(notepack.decode(notepack.encode('äß'))).to.equal('äß');
    // 2-byte
    expect(notepack.decode(notepack.encode('עִבְרִית'))).to.equal('עִבְרִית');
    // 3-byte
    expect(notepack.decode(notepack.encode('\u13DA'))).to.equal('\u13DA');
    // 4-byte
    expect(notepack.decode(notepack.encode('🌐'))).to.equal('🌐');
  });

  it('timestamp ext', function () {
    check(new Date(0), 'd6ff00000000');
    check(new Date('1956-06-17T00:00:00.000Z'), 'c70cff00000000ffffffffe6876500');
    check(new Date('1970-01-01T00:00:00.000Z'), 'd6ff00000000');
    check(new Date('2000-06-13T00:00:00.000Z'), 'd6ff39457980');
    check(new Date('2005-12-31T23:59:59.999Z'), 'd7ffee2e1f0043b71b7f');
    check(new Date('2140-01-01T13:14:15.678Z'), 'd7ffa1a5d6013fc2faa7');
  });

  it('all formats', function () {
    this.timeout(20000);
    const expected = {
      unsigned: [1, 2, 3, 4, { b: { c: [128, 256, 65536, 4294967296] } }],
      signed: [-1, -2, -3, -4, { b: { c: [-33, -129, -32769, -2147483649] } }],
      bin: [Uint8Array.of('1', '2', '3').buffer, Uint8Array.from('1'.repeat(256)).buffer, Uint8Array.from('2'.repeat(65536)).buffer],
      str: ['abc', 'g'.repeat(32), 'h'.repeat(256), 'i'.repeat(65536)],
      array: [[], array(16), array(65536)],
      map: {},
      nil: null,
      bool: { 'true': true, 'false': false, both: [true, false, false, false, true] },
      fixext: [null, new Date('2140-01-01T13:14:15.678Z'), null, new Date('2005-12-31T23:59:59.999Z')],
      utf8: ['α', '亜', '\uD83D\uDC26'],
      float: [1.1, 1234567891234567.5, Infinity, -Infinity, NaN]
    };
    expected.map['a'.repeat(32)] = { a: 'a', b: 'b', c: 'c' };
    expected.map['b'.repeat(256)] = { a: { b: 1, c: 1, d: 1, e: { f: { g: 2, h: 2 } } } };
    expected.map['c'.repeat(65536)] = [{ a: { b: 1, c: 1, d: 1, e: { f: [{ g: 2, h: 2 }] } } }];
    expected.map16 = map(65535);
    expected.map32 = map(65536);

    expect(notepack.decode(notepack.encode(expected))).to.deep.equal(expected);
  });
});


lib/encode.js
'use strict';

const MICRO_OPT_LEN = 32;
const TIMESTAMP32_MAX_SEC = 0x100000000 - 1; // 32-bit unsigned int
const TIMESTAMP64_MAX_SEC = 0x400000000 - 1; // 34-bit unsigned int

// Faster for short strings than buffer.write
function utf8Write(arr, offset, str) {
  let c = 0;
  for (let i = 0, l = str.length; i < l; i++) {
    c = str.charCodeAt(i);
    if (c < 0x80) {
      arr[offset++] = c;
    } else if (c < 0x800) {
      arr[offset++] = 0xc0 | (c >> 6);
      arr[offset++] = 0x80 | (c & 0x3f);
    } else if (c < 0xd800 || c >= 0xe000) {
      arr[offset++] = 0xe0 | (c >> 12);
      arr[offset++] = 0x80 | (c >> 6) & 0x3f;
      arr[offset++] = 0x80 | (c & 0x3f);
    } else {
      i++;
      c = 0x10000 + (((c & 0x3ff) << 10) | (str.charCodeAt(i) & 0x3ff));
      arr[offset++] = 0xf0 | (c >> 18);
      arr[offset++] = 0x80 | (c >> 12) & 0x3f;
      arr[offset++] = 0x80 | (c >> 6) & 0x3f;
      arr[offset++] = 0x80 | (c & 0x3f);
    }
  }
}

// Faster for short strings than Buffer.byteLength
function utf8Length(str) {
  let c = 0, length = 0;
  for (let i = 0, l = str.length; i < l; i++) {
    c = str.charCodeAt(i);
    if (c < 0x80) {
      length += 1;
    } else if (c < 0x800) {
      length += 2;
    } else if (c < 0xd800 || c >= 0xe000) {
      length += 3;
    } else {
      i++;
      length += 4;
    }
  }
  return length;
}

const cache = new Map();
const cacheMaxSize = process.env.NOTEPACK_ENCODE_CACHE_MAX_SIZE || 1024;

/*jshint latedef: nofunc */
function encodeKey(bytes, defers, key) {
  if (cache.has(key)) {
    const buffer = cache.get(key);
    defers.push({ bin: buffer, length: buffer.length, offset: bytes.length });
    return buffer.length;
  }
  if (cache.size > cacheMaxSize) {
    return _encode(bytes, defers, key);
  }
  const keyBytes = [];
  const size = _encode(keyBytes, [], key);
  const keyBuffer = Buffer.allocUnsafe(size);
  for (let i = 0, l = keyBytes.length; i < l; i++) {
    keyBuffer[i] = keyBytes[i];
  }
  utf8Write(keyBuffer, keyBytes.length, key);
  defers.push({ bin: keyBuffer, length: size, offset: bytes.length });
  cache.set(key, keyBuffer);
  return size;
}

function _encode(bytes, defers, value) {
  let hi = 0, lo = 0, length = 0, size = 0;

  switch (typeof value) {
    case 'string':
      if (value.length > MICRO_OPT_LEN) {
        length = Buffer.byteLength(value);
      } else {
        length = utf8Length(value);
      }

      if (length < 0x20) { // fixstr
        bytes.push(length | 0xa0);
        size = 1;
      } else if (length < 0x100) { // str 8
        bytes.push(0xd9, length);
        size = 2;
      } else if (length < 0x10000) { // str 16
        bytes.push(0xda, length >> 8, length);
        size = 3;
      } else if (length < 0x100000000) { // str 32
        bytes.push(0xdb, length >> 24, length >> 16, length >> 8, length);
        size = 5;
      } else {
        throw new Error('String too long');
      }
      defers.push({ str: value, length: length, offset: bytes.length });
      return size + length;
    case 'number':
      // TODO: encode to float 32?

      if (Math.floor(value) !== value || !isFinite(value)) { // float 64
        bytes.push(0xcb);
        defers.push({ float: value, length: 8, offset: bytes.length });
        return 9;
      }

      if (value >= 0) {
        if (value < 0x80) { // positive fixnum
          bytes.push(value);
          return 1;
        }

        if (value < 0x100) { // uint 8
          bytes.push(0xcc, value);
          return 2;
        }

        if (value < 0x10000) { // uint 16
          bytes.push(0xcd, value >> 8, value);
          return 3;
        }

        if (value < 0x100000000) { // uint 32
          bytes.push(0xce, value >> 24, value >> 16, value >> 8, value);
          return 5;
        }
        // uint 64
        hi = (value / Math.pow(2, 32)) >> 0;
        lo = value >>> 0;
        bytes.push(0xcf, hi >> 24, hi >> 16, hi >> 8, hi, lo >> 24, lo >> 16, lo >> 8, lo);
        return 9;
      } else {

        if (value >= -0x20) { // negative fixnum
          bytes.push(value);
          return 1;
        }

        if (value >= -0x80) { // int 8
          bytes.push(0xd0, value);
          return 2;
        }

        if (value >= -0x8000) { // int 16
          bytes.push(0xd1, value >> 8, value);
          return 3;
        }

        if (value >= -0x80000000) { // int 32
          bytes.push(0xd2, value >> 24, value >> 16, value >> 8, value);
          return 5;
        }
        // int 64
        hi = Math.floor(value / Math.pow(2, 32));
        lo = value >>> 0;
        bytes.push(0xd3, hi >> 24, hi >> 16, hi >> 8, hi, lo >> 24, lo >> 16, lo >> 8, lo);
        return 9;
      }
      break;
    case 'object':
      // nil
      if (value === null) {
        bytes.push(0xc0);
        return 1;
      }

      if (Array.isArray(value)) {
        length = value.length;

        if (length < 0x10) { // fixarray
          bytes.push(length | 0x90);
          size = 1;
        } else if (length < 0x10000) { // array 16
          bytes.push(0xdc, length >> 8, length);
          size = 3;
        } else if (length < 0x100000000) { // array 32
          bytes.push(0xdd, length >> 24, length >> 16, length >> 8, length);
          size = 5;
        } else {
          throw new Error('Array too large');
        }
        for (let i = 0; i < length; i++) {
          size += _encode(bytes, defers, value[i]);
        }
        return size;
      }

      if (value instanceof Date) {
        const ms = value.getTime();
        const s = Math.floor(ms / 1e3);
        const ns = (ms - s * 1e3) * 1e6;

        if (s >= 0 && ns >= 0 && s <= TIMESTAMP64_MAX_SEC) {
          if (ns === 0 && s <= TIMESTAMP32_MAX_SEC) {
            // timestamp 32
            bytes.push(0xd6, 0xff, s >> 24, s >> 16, s >> 8, s);
            return 6;
          } else {
            // timestamp 64
            hi = s / 0x100000000;
            lo = s & 0xffffffff;
            bytes.push(0xd7, 0xff, ns >> 22, ns >> 14, ns >> 6, hi, lo >> 24, lo >> 16, lo >> 8, lo);
            return 10;
          }
        } else {
          // timestamp 96
          hi = Math.floor(s / 0x100000000);
          lo = s >>> 0;
          bytes.push(0xc7, 0x0c, 0xff, ns >> 24, ns >> 16, ns >> 8, ns, hi >> 24, hi >> 16, hi >> 8, hi, lo >> 24, lo >> 16, lo >> 8, lo);
          return 15;
        }
      }

      if (value instanceof Buffer) {
        length = value.length;

        if (length < 0x100) { // bin 8
          bytes.push(0xc4, length);
          size = 2;
        } else if (length < 0x10000) { // bin 16
          bytes.push(0xc5, length >> 8, length);
          size = 3;
        } else if (length < 0x100000000) { // bin 32
          bytes.push(0xc6, length >> 24, length >> 16, length >> 8, length);
          size = 5;
        } else {
          throw new Error('Buffer too large');
        }
        defers.push({ bin: value, length: length, offset: bytes.length });
        return size + length;
      }

      if (value instanceof ArrayBuffer || ArrayBuffer.isView(value)) {
        const arraybuffer = value.buffer || value;
        length = arraybuffer.byteLength;

        // ext 8
        if (length < 0x100) {
          bytes.push(0xc7, length, 0);
          size = 3;
        } else if (length < 0x10000) { // ext 16
          bytes.push(0xc8, length >> 8, length, 0);
          size = 4;
        } else if (length < 0x100000000) { // ext 32
          bytes.push(0xc9, length >> 24, length >> 16, length >> 8, length, 0);
          size = 6;
        } else {
          throw new Error('ArrayBuffer too large');
        }
        defers.push({ arraybuffer: arraybuffer, length: length, offset: bytes.length });
        return size + length;
      }

      if (typeof value.toJSON === 'function') {
        return _encode(bytes, defers, value.toJSON());
      }

      // note: faster than for...in + hasOwnProperty in most cases
      const keys = [], allKeys = Object.keys(value);
      let key = '';

      for (let i = 0, l = allKeys.length; i < l; i++) {
        key = allKeys[i];
        if (value[key] !== undefined && typeof value[key] !== 'function') {
          keys.push(key);
        }
      }
      length = keys.length;

      if (length < 0x10) { // fixmap
        bytes.push(length | 0x80);
        size = 1;
      } else if (length < 0x10000) { // map 16
        bytes.push(0xde, length >> 8, length);
        size = 3;
      } else if (length < 0x100000000) { // map 32
        bytes.push(0xdf, length >> 24, length >> 16, length >> 8, length);
        size = 5;
      } else {
        throw new Error('Object too large');
      }

      for (let i = 0; i < length; i++) {
        key = keys[i];
        size += encodeKey(bytes, defers, key);
        size += _encode(bytes, defers, value[key]);
      }
      return size;
    case 'boolean': // false/true
      bytes.push(value ? 0xc3 : 0xc2);
      return 1;
    case 'undefined':
      bytes.push(0xc0);
      return 1;
    default:
      // custom types like BigInt (typeof value === 'bigint')
      if (typeof value.toJSON === 'function') {
        return _encode(bytes, defers, value.toJSON());
      }
      throw new Error('Could not encode');
  }
}

function encode(value) {
  const bytes = [], defers = [], size = _encode(bytes, defers, value);
  const buf = Buffer.allocUnsafe(size);

  let deferIndex = 0, deferWritten = 0, nextOffset = -1;
  if (defers.length > 0) {
    nextOffset = defers[0].offset;
  }

  let defer, deferLength = 0, offset = 0;
  for (let i = 0, l = bytes.length; i < l; i++) {
    buf[deferWritten + i] = bytes[i];
    while (i + 1 === nextOffset) {
      defer = defers[deferIndex];
      deferLength = defer.length;
      offset = deferWritten + nextOffset;
      if (defer.bin) {
        if (deferLength > MICRO_OPT_LEN) {
          defer.bin.copy(buf, offset, 0, deferLength);
        } else {
          const bin = defer.bin;
          for (let j = 0; j < deferLength; j++) {
            buf[offset + j] = bin[j];
          }
        }
      } else if (defer.str) {
        if (deferLength > MICRO_OPT_LEN) {
          buf.write(defer.str, offset, deferLength, 'utf8');
        } else {
          utf8Write(buf, offset, defer.str);
        }
      } else if (defer.float !== undefined) {
        buf.writeDoubleBE(defer.float, offset);
      } else if (defer.arraybuffer) {
        const arr = new Uint8Array(defer.arraybuffer);
        for (let k = 0; k < deferLength; k++) {
          buf[offset + k] = arr[k];
        }
      }
      deferIndex++;
      deferWritten += deferLength;
      if (defers[deferIndex]) {
        nextOffset = defers[deferIndex].offset;
      } else {
        break;
      }
    }
  }
  return buf;
}

module.exports = encode;
